### ISSUE TITLE
Sharing loss function code

### ISSUE THREAD

**Issue by palmex** [IDEA-Research/MotionLLM#13]
Are you able to share the loss function code or refer us to the source which you are utilizing?

**Comment by LinghaoChan** [comment]
@palmex we follow the lit-gpt codebase. You can read [this](https://github.com/Lightning-AI/litgpt/blob/main/tutorials/finetune_lora.md).

**Comment by palmex** [comment]
Are you pooling the text and motion embeddings to have the same length before applying contrastive loss?

**Comment by LinghaoChan** [comment]
> Are you pooling the text and motion embeddings to have the same length before applying contrastive loss?

Sorry, I do not get your point. Why you need this~~

**Comment by palmex** [comment]
I'm trying to create my own training script, and I do not have sufficient details from your paper. 
Are you using contrastive loss from a specific library, like CLIP or MotionCLIP? I'm assuming you are using token-level contrastive loss.

**Comment by LinghaoChan** [comment]
> I'm trying to create my own training script, and I do not have sufficient details from your paper. Are you using contrastive loss from a specific library, like CLIP or MotionCLIP? I'm assuming you are using token-level contrastive loss.

Our training loss is the CE loss by token, not the contrastive loss.

**Comment by palmex** [comment]
I think I understand better now. 
Stage 1: 
Projected Motion Tokens (input) => Generated LLM Text Tokens (output)
Loss = CE(motion caption (gt) vs. text output (pred))

I notice that the model accepts a tuple as input of (text,motion,text). For Stage 1: is there any text passed into the LLM along with the motion tokens?

**Comment by LinghaoChan** [comment]
> I think I understand better now. Stage 1: Projected Motion Tokens (input) => Generated LLM Text Tokens (output) Loss = CE(motion caption (gt) vs. text output (pred))
> 
> I notice that the model accepts a tuple as input of (text,motion,text). For Stage 1: is there any text passed into the LLM along with the motion tokens?

@palmex  Yes, you are right. The input is (text,motion,text), aka. (system command, motion, Q-A).

**Comment by palmex** [comment]
What were your loss values after Stage1? How many of the motion caption tokens were you able to predict correctly? 

Also, your paper mentions 40k iterations: is this 40k epochs? Is your batch size = 1? 

This is on the motion translation only, and the model is able to predict up to 50% of the tokens correctly (so far).

![Screenshot 2024-12-11 at 11 13 56 AM](https://github.com/user-attachments/assets/b60513e5-055b-45b5-a82c-f1f26ba393b2)

**Comment by LinghaoChan** [comment]
I remember the converged loss is about 0.7-0.9 at stage 1. 

40 iters, not epochs. I remember the bs=4 or 8.

I think the caption result of Stage 1 is good enough.
